File: src/C3.Infrastructure/Updates/UpdateCheckSchedule.vb
Namespace Updates

    Public NotInheritable Class UpdateCheckSchedule

        Private Sub New()
        End Sub

        Public Shared Function Parse(value As String) As UpdateCheckPolicy
            Dim policy As UpdateCheckPolicy
            If TryParseStored(value, policy) Then
                Return policy
            End If
            Return UpdateCheckPolicy.Never
        End Function

        Public Shared Function TryParseStored(
                value As String,
                ByRef policy As UpdateCheckPolicy) As Boolean

            Select Case If(value, String.Empty).Trim().ToLowerInvariant()
                Case "true", "startup"
                    policy = UpdateCheckPolicy.Startup
                    Return True
                Case "weekly"
                    policy = UpdateCheckPolicy.Weekly
                    Return True
                Case "monthly"
                    policy = UpdateCheckPolicy.Monthly
                    Return True
                Case "false", "manually", "never"
                    policy = UpdateCheckPolicy.Never
                    Return True
                Case Else
                    policy = UpdateCheckPolicy.Never
                    Return False
            End Select
        End Function

        Public Shared Function Serialize(value As UpdateCheckPolicy) As String
            Select Case value
                Case UpdateCheckPolicy.Startup
                    Return "startup"
                Case UpdateCheckPolicy.Weekly
                    Return "weekly"
                Case UpdateCheckPolicy.Monthly
                    Return "monthly"
                Case Else
                    Return "never"
            End Select
        End Function

        Public Shared Function ShouldCheck(
                policy As UpdateCheckPolicy,
                lastCheckedAt As DateTime,
                now As DateTime) As Boolean

            If policy = UpdateCheckPolicy.Never OrElse
                    Not [Enum].IsDefined(GetType(UpdateCheckPolicy), policy) Then
                Return False
            End If
            Dim normalizedNow As DateTime = NormalizeUtc(now)
            Dim normalizedLastCheck As DateTime = If(
                lastCheckedAt = DateTime.MinValue,
                DateTime.MinValue,
                NormalizeUtc(lastCheckedAt))
            If normalizedLastCheck <> DateTime.MinValue AndAlso
                    (normalizedLastCheck - normalizedNow).TotalMinutes > 5D Then
                Return True
            End If

            Select Case policy
                Case UpdateCheckPolicy.Startup
                    Return True
                Case UpdateCheckPolicy.Weekly
                    Return normalizedLastCheck = DateTime.MinValue OrElse
                        (normalizedNow - normalizedLastCheck).TotalDays >= 7D
                Case UpdateCheckPolicy.Monthly
                    Return normalizedLastCheck = DateTime.MinValue OrElse
                        (normalizedNow - normalizedLastCheck).TotalDays >= 28D
                Case Else
                    Return False
            End Select
        End Function

        Public Shared Function NormalizeUtc(value As DateTime) As DateTime
            If value = DateTime.MinValue OrElse value.Kind = DateTimeKind.Utc Then
                Return value
            End If
            If value.Kind = DateTimeKind.Unspecified Then
                value = DateTime.SpecifyKind(value, DateTimeKind.Local)
            End If
            Return value.ToUniversalTime()
        End Function

    End Class

End Namespace

File: src/C3.Infrastructure/Updates/UpdateCheckSchedule.cs
using Microsoft.VisualBasic.CompilerServices;
using System;

namespace C3.Infrastructure.Updates
{
	public sealed class UpdateCheckSchedule
	{
		private UpdateCheckSchedule()
		{
		}

		public static UpdateCheckPolicy Parse(string value)
		{
			UpdateCheckPolicy result = default(UpdateCheckPolicy);
			if (TryParseStored(value, ref result))
			{
				return result;
			}
			return UpdateCheckPolicy.Never;
		}

		public static bool TryParseStored(string value, ref UpdateCheckPolicy policy)
		{
			string text = (value ?? string.Empty).Trim().ToLowerInvariant();
			switch (text)
			{
			case "true":
			case "startup":
				policy = UpdateCheckPolicy.Startup;
				return true;
			case "weekly":
				policy = UpdateCheckPolicy.Weekly;
				return true;
			case "monthly":
				policy = UpdateCheckPolicy.Monthly;
				return true;
			case "false":
			case "manually":
			case "never":
				policy = UpdateCheckPolicy.Never;
				return true;
			default:
				policy = UpdateCheckPolicy.Never;
				return false;
			}
		}

		public static string Serialize(UpdateCheckPolicy value)
		{
			switch (value)
			{
			case UpdateCheckPolicy.Startup:
				return "startup";
			case UpdateCheckPolicy.Weekly:
				return "weekly";
			case UpdateCheckPolicy.Monthly:
				return "monthly";
			default:
				return "never";
			}
		}

		public static bool ShouldCheck(UpdateCheckPolicy policy, DateTime lastCheckedAt, DateTime now)
		{
			if (policy != 0 && Enum.IsDefined(typeof(UpdateCheckPolicy), policy))
			{
				DateTime dateTime = NormalizeUtc(now);
				DateTime dateTime2 = (DateTime.Compare(lastCheckedAt, DateTime.MinValue) == 0) ? DateTime.MinValue : NormalizeUtc(lastCheckedAt);
				TimeSpan timeSpan;
				if (DateTime.Compare(dateTime2, DateTime.MinValue) != 0)
				{
					timeSpan = dateTime2 - dateTime;
					if (timeSpan.TotalMinutes > 5.0)
					{
						return true;
					}
				}
				switch (policy)
				{
				case UpdateCheckPolicy.Startup:
					return true;
				case UpdateCheckPolicy.Weekly:
				{
					int result2;
					if (DateTime.Compare(dateTime2, DateTime.MinValue) != 0)
					{
						timeSpan = dateTime - dateTime2;
						result2 = ((timeSpan.TotalDays >= 7.0) ? 1 : 0);
					}
					else
					{
						result2 = 1;
					}
					return (byte)result2 != 0;
				}
				case UpdateCheckPolicy.Monthly:
				{
					int result;
					if (DateTime.Compare(dateTime2, DateTime.MinValue) != 0)
					{
						timeSpan = dateTime - dateTime2;
						result = ((timeSpan.TotalDays >= 28.0) ? 1 : 0);
					}
					else
					{
						result = 1;
					}
					return (byte)result != 0;
				}
				default:
					return false;
				}
			}
			return false;
		}

		public static DateTime NormalizeUtc(DateTime value)
		{
			if (DateTime.Compare(value, DateTime.MinValue) != 0 && value.Kind != DateTimeKind.Utc)
			{
				if (value.Kind == DateTimeKind.Unspecified)
				{
					value = DateTime.SpecifyKind(value, DateTimeKind.Local);
				}
				return value.ToUniversalTime();
			}
			return value;
		}
	}
}
